File: src/lib/fees.ts
/**
 * Floors on the three amounts a player can choose, in USD.
 *
 * Lowered from $10 so a newer platform is not asking for a $10 commitment
 * before anyone has seen a match run. They move together on purpose: a deposit
 * floor above the entry floor leaves money that cannot be staked, and a
 * withdrawal floor above either strands a player who deposited the minimum,
 * played it, and then cannot take their winnings out.
 *
 * Card processing is the reason these have a floor at all — Stripe takes
 * roughly 2.9% + 30c, so a $5 deposit costs about 45c before anyone plays. The
 * 10% fee on the resulting $10 pool covers it, but the margin is thinner than
 * it was at $10, which is the trade for the lower barrier.
 */
export const MIN_ENTRY_USD = 5;
export const MIN_DEPOSIT_USD = 5;
export const MIN_WITHDRAWAL_USD = 5;

/**
 * Supported games.
 *
 * Slugs are stable identifiers stored in the database (`challenges.game_slug`,
 * `tournaments.game_slug`), so they deliberately do NOT carry a year — only the
 * display labels do. That way next season's rename is a one-line label change
 * and never a data migration.
 */
export const SUPPORTED_GAMES = ["fortnite", "nba2k", "madden", "ncaa", "mlbshow"] as const;
export type SupportedGame = (typeof SUPPORTED_GAMES)[number];

/**
 * Platforms a match can be played on.
 *
 * "Crossplay" is not a console — it is the player saying the platform does not
 * matter, which most of these titles now support. It sits in the same list
 * because that is the one question the challenge form asks, and splitting it
 * into a separate toggle would make every existing challenge ambiguous about
 * whether it had opted in.
 *
 * Stored as free text on `challenges.platform`, so adding one is a code change
 * and never a migration. Kept here rather than in each route because two copies
 * had already drifted apart once.
 */
export const PLATFORMS = ["PC", "PlayStation", "Xbox", "Switch", "Crossplay"] as const;
export type Platform = (typeof PLATFORMS)[number];

export const GAME_LABELS: Record<SupportedGame, string> = {
  fortnite: "Fortnite",
  nba2k: "NBA 2K27",
  madden: "Madden NFL 27",
  ncaa: "NCAA 27",
  mlbshow: "MLB The Show 26",
};

/**
 * MatchPoint tiered service fee.
 *
 * Fee is calculated against the TOTAL prize pool (sum of all entry fees).
 * Lower rates on bigger pools to incentivize larger events.
 *
 *   Pool size         Rate
 *   $1    – $25       10%
 *   $26   – $100       8%
 *   $101  – $500       6%
 *   $501+              5%
 */

export type FeeTier = {
  minPool: number;
  maxPool: number; // inclusive upper bound; Infinity for top tier
  rate: number; // 0..1
  label: string;
};

export const FEE_TIERS: ReadonlyArray<FeeTier> = [
  { minPool: 0, maxPool: 25, rate: 0.1, label: "$1 – $25" },
  { minPool: 25.01, maxPool: 100, rate: 0.08, label: "$26 – $100" },
  { minPool: 100.01, maxPool: 500, rate: 0.06, label: "$101 – $500" },
  { minPool: 500.01, maxPool: Infinity, rate: 0.05, label: "$501+" },
];

export type FeeBreakdown = {
  pool: number;
  rate: number;
  tierLabel: string;
  serviceFee: number;
  netPrize: number;
};

function round2(n: number) {
  return Math.round(n * 100) / 100;
}

export function getFeeTier(pool: number): FeeTier {
  const p = Math.max(0, Number(pool) || 0);
  for (const t of FEE_TIERS) {
    if (p >= t.minPool && p <= t.maxPool) return t;
  }
  return FEE_TIERS[FEE_TIERS.length - 1];
}

export function getFeeRate(pool: number): number {
  return getFeeTier(pool).rate;
}

/**
 * Split a pool into service fee + net prize using the tiered rate.
 * Used for escrow payout: keep `serviceFee`, transfer `netPrize` to winner.
 */
export function calculateFee(pool: number): FeeBreakdown {
  const p = round2(Math.max(0, Number(pool) || 0));
  const tier = getFeeTier(p);
  const serviceFee = round2(p * tier.rate);
  const netPrize = round2(p - serviceFee);
  return { pool: p, rate: tier.rate, tierLabel: tier.label, serviceFee, netPrize };
}

/** Tournament: pool = entryFee * playerCount. */
export function calculateTournamentFee(entryFee: number, playerCount: number): FeeBreakdown {
  const pool = round2(
    Math.max(0, Number(entryFee) || 0) * Math.max(0, Math.floor(Number(playerCount) || 0)),
  );
  return calculateFee(pool);
}

/** 1v1 challenge: pool = entryAmount * 2 (both players pay). */
export function calculateChallengeFee(entryAmount: number): FeeBreakdown {
  const pool = round2(Math.max(0, Number(entryAmount) || 0) * 2);
  return calculateFee(pool);
}

/* =========================================================================
 *  Withdrawal fees
 *
 *  Standard payouts (2–5 business days) are FREE.
 *  Same-day payouts cost a flat 8% of the amount withdrawn — no tiers, no
 *  minimum, the same rate at $10 as at $5,000.
 *
 *  Flat on purpose. Kevin asked for percentage-based pricing after tiered
 *  flat fees, and a single rate is the version a player can check in their
 *  head before they tap the button. A tier table only earns its complexity if
 *  the rate actually changes across it.
 * ========================================================================= */

export type WithdrawalSpeed = "standard" | "same_day";

/** Same-day withdrawal fee, as a fraction of the gross amount. */
export const SAME_DAY_WITHDRAWAL_RATE = 0.08;

export type WithdrawalFeeBreakdown = {
  speed: WithdrawalSpeed;
  grossCents: number;
  feeCents: number;
  netCents: number;
  /** Percentage rate charged (0..1); 0 for standard payouts. */
  rate: number;
  tierLabel: string;
  etaLabel: string;
};

export function calculateWithdrawalFee(
  amountCents: number,
  speed: WithdrawalSpeed,
): WithdrawalFeeBreakdown {
  const gross = Math.max(0, Math.floor(Number(amountCents) || 0));
  if (speed === "standard") {
    return {
      speed,
      grossCents: gross,
      feeCents: 0,
      netCents: gross,
      rate: 0,
      tierLabel: "Free",
      etaLabel: "2–5 business days",
    };
  }
  // Capped at the withdrawal itself so a rounding edge can never return a
  // negative net.
  const fee = Math.min(gross, Math.round(gross * SAME_DAY_WITHDRAWAL_RATE));
  return {
    speed,
    grossCents: gross,
    feeCents: fee,
    netCents: Math.max(0, gross - fee),
    rate: SAME_DAY_WITHDRAWAL_RATE,
    tierLabel: "8% of the amount",
    etaLabel: "Typically 30 minutes – 5 hours",
  };
}

File: backend/fees.py
"""MatchPoint tiered fees.

Platform service fee is charged against the TOTAL prize pool (sum of all
entry fees) with lower rates on bigger pools to incentivize larger events.

    Pool size          Rate
    $1     - $25       10%
    $26    - $100       8%
    $101   - $500       6%
    $501+               5%

Withdrawal fees:
    Standard (2-5 business days): FREE
    Same-day (30 min - 5 hours), amount tiers:
        $10   - $50      $1.99
        $51   - $100     $2.99
        $101  - $250     $4.99
        $251  - $500     $7.99
        $501  - $1,000   $12.99
        $1,001+          1% of amount
"""
from dataclasses import dataclass
from typing import List, Literal


@dataclass(frozen=True)
class FeeTier:
    min_pool: float
    max_pool: float  # inclusive; math.inf for top tier
    rate: float
    label: str


FEE_TIERS: List[FeeTier] = [
    FeeTier(0.0, 25.0, 0.10, "$1 – $25"),
    FeeTier(25.01, 100.0, 0.08, "$26 – $100"),
    FeeTier(100.01, 500.0, 0.06, "$101 – $500"),
    FeeTier(500.01, float("inf"), 0.05, "$501+"),
]


def _round2(n: float) -> float:
    return round(n * 100) / 100


def get_fee_tier(pool: float) -> FeeTier:
    p = max(0.0, float(pool or 0))
    for t in FEE_TIERS:
        if t.min_pool <= p <= t.max_pool:
            return t
    return FEE_TIERS[-1]


def get_fee_rate(pool: float) -> float:
    return get_fee_tier(pool).rate


@dataclass
class FeeBreakdown:
    pool: float
    rate: float
    tier_label: str
    service_fee: float
    net_prize: float


def calculate_fee(pool: float) -> FeeBreakdown:
    p = _round2(max(0.0, float(pool or 0)))
    tier = get_fee_tier(p)
    service_fee = _round2(p * tier.rate)
    net_prize = _round2(p - service_fee)
    return FeeBreakdown(pool=p, rate=tier.rate, tier_label=tier.label,
                        service_fee=service_fee, net_prize=net_prize)


def calculate_tournament_fee(entry_fee: float, player_count: int) -> FeeBreakdown:
    pool = _round2(max(0.0, float(entry_fee or 0)) * max(0, int(player_count or 0)))
    return calculate_fee(pool)


def calculate_challenge_fee(entry_amount: float) -> FeeBreakdown:
    """1v1 pool = stake * 2."""
    pool = _round2(max(0.0, float(entry_amount or 0)) * 2)
    return calculate_fee(pool)


# ------------------- Withdrawal tiers -------------------
WithdrawalSpeed = Literal["standard", "same_day"]


@dataclass(frozen=True)
class WithdrawalTier:
    min_cents: int
    max_cents: int  # inclusive; sys.maxsize for top tier
    flat_fee_cents: int  # 0 when using pct_rate
    pct_rate: float  # 0 when using flat
    label: str


SAME_DAY_WITHDRAWAL_TIERS: List[WithdrawalTier] = [
    WithdrawalTier(1000, 5000, 199, 0.0, "$10 – $50"),
    WithdrawalTier(5001, 10000, 299, 0.0, "$51 – $100"),
    WithdrawalTier(10001, 25000, 499, 0.0, "$101 – $250"),
    WithdrawalTier(25001, 50000, 799, 0.0, "$251 – $500"),
    WithdrawalTier(50001, 100000, 1299, 0.0, "$501 – $1,000"),
    WithdrawalTier(100001, 10**12, 0, 0.01, "$1,001+"),
]


@dataclass
class WithdrawalFeeBreakdown:
    speed: str
    gross_cents: int
    fee_cents: int
    net_cents: int
    tier_label: str
    eta_label: str


def calculate_withdrawal_fee(amount_cents: int, speed: str) -> WithdrawalFeeBreakdown:
    gross = max(0, int(amount_cents or 0))
    if speed == "standard":
        return WithdrawalFeeBreakdown(speed="standard", gross_cents=gross,
                                      fee_cents=0, net_cents=gross,
                                      tier_label="Free",
                                      eta_label="2–5 business days")
    tier = next((t for t in SAME_DAY_WITHDRAWAL_TIERS if t.min_cents <= gross <= t.max_cents),
                SAME_DAY_WITHDRAWAL_TIERS[-1])
    if tier.flat_fee_cents > 0:
        fee = tier.flat_fee_cents
    else:
        fee = max(1, round(gross * tier.pct_rate))
    net = max(0, gross - fee)
    return WithdrawalFeeBreakdown(speed="same_day", gross_cents=gross,
                                  fee_cents=fee, net_cents=net,
                                  tier_label=tier.label,
                                  eta_label="Typically 30 minutes – 5 hours")
